File: Easy/product_sum.py
def productSum(array):
    return productSumHelper(array, depth=1)


def productSumHelper(array, depth=1):
    sum = 0

    for num in array:
        if isinstance(num, list):
            sum += productSumHelper(num, depth + 1)
        else:
            sum += num

    return sum * depth

# Time Complexity: O(n)
# Space Complexity: O(d) where d is the greatest depth in the array

File: Javascript/Arrays/product_sum.js
function productSum(array, depth = 1) {
	let sum = 0;

	for (let i = 0; i < array.length; i++) {
		const num = array[i];

		if (Array.isArray(num)) {
			sum += productSum(num, depth + 1);
		} else {
			sum += num;
		}
	}
	return sum * depth;
}

// Time Complexity - O(n)
// Space Complexity - O(d)
